### PR TITLE
[semaphore] implement a `Semaphore.call_linksafe/3`.

### PR DESCRIPTION
**Pull request by jhgg**
Attempts to address #2 

This function will automatically release the semaphore if the process
dies in such a way that cannot be caught in the `after` block, e.g.
a linked process dies during the `func` callback.

This is a bit of an RFC before I write tests around it...

### CODE REVIEW

File: lib/semaphore.ex [hunk 1]
```diff
@@ -83,9 +89,61 @@ defmodule Semaphore do
     end
   end
 
+  @doc """
+  Attempt to acquire a semaphore and call a function that might link to another process, and then automatically release.
+
+  If the current process dies in a way that is unable to be caught by the try block (e.g. a linked process dies, while
+  `func` is being called. The semaphore will be automatically released by the sweeper in the background.
+
+  This function has higher overhead than `call/3` and should only be used if you know that you might be linking to
```
Comment (scohen) [inline lib/semaphore.ex]: Is the overhead due to the extra `ETS` insert?

File: lib/semaphore.ex [hunk 2]
```diff
@@ -83,9 +89,61 @@ defmodule Semaphore do
...
+      end
+    else
+      {:error, :max}
+    end
+  end
+
   ## Private
 
-  defp init do
+  def init(sweep_interval) do
```
Comment (scohen) [inline lib/semaphore.ex]: This is so funny, I was just drafting a reply to Stan's comment last night suggesting this approach.

File: lib/semaphore.ex [hunk 3]
```diff
@@ -83,9 +89,61 @@ defmodule Semaphore do
...
+    do_sweep()
+    {:noreply, sweep_interval, sweep_interval}
+  end
+
+  defp do_sweep() do
+    ETS.foldl(
+      fn ({name, pid} = key, :ok) ->
+        with false <- Process.alive?(pid),
+             1 <- :ets.select_delete(@call_safe_table, [{key, [], [true]}]) do
+          IO.puts("releasing #{inspect key}")
```
Comment (scohen) [inline lib/semaphore.ex]: `Logger.warn` ?

File: lib/semaphore.ex (hunk 1 above)
Comment (jhgg) [reply lib/semaphore.ex]: Yeah. The extra insert & delete. It's 2x the ETS calls. Additionally, it makes the sweeper do more work.

File: lib/semaphore.ex (hunk 3 above)
Comment (jhgg) [reply lib/semaphore.ex]: This was just for my testing. Gonna remove it once I finalize this approach.

**Pull request merged**